### PR TITLE
Features/sma 233 gas opt 1 + 2 + 3

### PR DESCRIPTION
**Pull request by livingrockrises**
# Summary

This also includes following merged changes

https://github.com/bcnmy/biconomy-paymasters/pull/60
https://github.com/bcnmy/biconomy-paymasters/pull/63
https://github.com/bcnmy/biconomy-paymasters/pull/67
https://github.com/bcnmy/biconomy-paymasters/pull/69

**result (after)**

paymaster and data length 278
// without oracle, with mock oracle, with real oracle on testnet
validatePaymasterUserOp():
postOp():

Mumbai with real oracle
[14107] BiconomyTokenPaymaster::validatePaymasterUserOp
[56554] BiconomyTokenPaymaster::postOp
[178173] Entry Point::handleOps

**Before:**

pnd length  558

// without oracle, with mock oracle, with real oracle on testnet
validatePaymasterUserOp()
postOp()

Mumbai with real oracle
[15322] BiconomyTokenPaymaster::validatePaymasterUserOp
[72271] BiconomyTokenPaymaster::postOp
[195432] Entry Point::handleOps

**Total Savings**

(72271 - 56554) + (15322 - 14107) = 16932 + calldata reduction cost

[ handOps diff 17259]

Things left:

- Add more accounting tests
- Update deployment scripts and deployment config
- savings comparison for tokens for which we don't use gas oracles

Related Issue: # (issue number)

## Change Type

- [ ] Bug Fix
- [ ] Refactor
- [ ] New Feature
- [ ] Breaking Change
- [ ] Documentation Update
- [ ] Performance Improvement
- [ ] Other

# Checklist

- [ ] My code follows this project's style guidelines
- [ ] I've reviewed my own code
- [ ] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [ ] My changes generate no new warnings
- [ ] I've added tests that prove my fix is effective or my feature works
- [ ] All unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

# Additional Information

# Branch Naming

### CODE REVIEW

File: contracts/token/oracles/OracleAggregator.sol [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    error LatestRoundIncomplete();
+    error PriceFeedStale();
+    error OracleAddressCannotBeZero();
+
+     struct TokenInfo {
+        uint8 tokenDecimals;
+        address tokenOracle;
+        address nativeOracle;
+        bool isDerivedFeed;
+     }
```
Comment (ankurdubey521) [inline contracts/token/oracles/OracleAggregator.sol:21]: errors, events and structs should be moved to interface

File: contracts/token/oracles/OracleAggregator.sol [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        tokensInfo[token].tokenOracle = tokenOracle;
+        tokensInfo[token].nativeOracle = nativeOracle;
+        tokensInfo[token].tokenDecimals = tokenDecimals;
+        tokensInfo[token].isDerivedFeed = isDerivedFeed;
+    }
+
+    /**
+     * @dev exchangeRate : each aggregator implements this method based on how it sources the quote/price
+     * @notice here it is token / native sourced from chainlink so in order to get defined exchangeRate we inverse the feed
+     * @param token ERC20 token address
```
Comment (ankurdubey521) [inline contracts/token/oracles/OracleAggregator.sol:54]: natspec does not include return value

File: contracts/token/oracles/OracleAggregator.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+             tokenPrice = 
+                fetchPrice(FeedInterface(tokenInfo.tokenOracle));
+             tokenOracleDecimals = FeedInterface(tokenInfo.tokenOracle).decimals();
+        }
+    }
+
+     /// @notice Fetches the latest price from the given Oracle.
+    /// @dev This function is used to get the latest price from the tokenOracle or nativeOracle.
+    /// @param _oracle The Oracle contract to fetch the price from.
+    /// @return price The latest price fetched from the Oracle.
```
Comment (ankurdubey521) [inline contracts/token/oracles/OracleAggregator.sol:95]: inconsistent natspec format
/// vs /** */

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 4]
```diff
@@ -8,13 +8,14 @@ import {UserOperation} from "@account-abstraction/contracts/interfaces/UserOpera
...
 import {BasePaymaster} from "../BasePaymaster.sol";
-import {IOracleAggregator} from "./oracles/IOracleAggregator.sol";
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@account-abstraction/contracts/core/Helpers.sol" as Helpers;
 import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
 import "../utils/SafeTransferLib.sol";
 import {TokenPaymasterErrors} from "./TokenPaymasterErrors.sol";
 import "@openzeppelin/contracts/utils/Address.sol";
+import {OracleAggregator} from "./oracles/OracleAggregator.sol";
+// import "hardhat/console.sol";
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:18]: Remove commented out import

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 5]
```diff
@@ -113,13 +114,21 @@ contract BiconomyTokenPaymaster is
...
+    // Review can replace with this event // can add exchangeRate
+    // saves 1000 gas
+    /*event UserOperationSponsored(
+        address indexed sender,
+        address indexed token,
+        uint256 indexed totalCharge,
+        uint256 actualGasCost
+    );*/
+
     /**
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:132]: remove commented out event

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 6]
```diff
@@ -113,13 +114,21 @@ contract BiconomyTokenPaymaster is
         address indexed sender,
         address indexed token,
         uint256 indexed totalCharge,
-        address oracleAggregator,
         uint32 priceMarkup,
         bytes32 userOpHash,
-        uint256 exchangeRate,
+        uint128 exchangeRate,
         ExchangeRateSource priceSource
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:120]: Can we please move these to the interface?

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 7]
```diff
@@ -326,6 +315,18 @@ contract BiconomyTokenPaymaster is
...
+        bytes calldata data
+    ) internal pure returns (bytes32 ret) {
+        assembly ("memory-safe") {
+            let mem := mload(0x40)
+            let len := data.length
+            calldatacopy(mem, data.offset, len)
+            ret := keccak256(mem, len) 
+        }
+    }
+
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:329]: let's move it to the helper's library

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 8]
```diff
@@ -395,17 +393,11 @@ contract BiconomyTokenPaymaster is
...
-            priceMarkup
-        ) = abi.decode(
-            paymasterAndData[VALID_PND_OFFSET:SIGNATURE_OFFSET],
-            (uint48, uint48, address, address, uint256, uint32)
-        );
+        validUntil = uint48(bytes6(paymasterAndData[21:27]));
+        validAfter = uint48(bytes6(paymasterAndData[27:33]));
+        feeToken = address(bytes20(paymasterAndData[33:53]));
+        exchangeRate = uint128(bytes16(paymasterAndData[53:69]));
+        priceMarkup = uint32(bytes4(paymasterAndData[69:73]));
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:400]: Do you want to parameterize the constants?

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 9]
```diff
@@ -443,43 +435,30 @@ contract BiconomyTokenPaymaster is
...
         (requiredPreFund);
-        // verificationGasLimit is dual-purposed, as gas limit for postOp. make sure it is high enough
-        // make sure that verificationGasLimit is high enough to handle postOp
-        require(
-            userOp.verificationGasLimit > UNACCOUNTED_COST,
-            "BTPM: gas too low for postOp"
-        );
 
-        // review: in this method try to resolve stack too deep (though via-ir is good enough)
+        // uint256 gas = gasleft();
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:439]: should be removed

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 10]
```diff
@@ -443,43 +435,30 @@ contract BiconomyTokenPaymaster is
...
         ) = parsePaymasterAndData(userOp.paymasterAndData);
+        // console.log("gas used for parsePmd: %s", gas - gasleft());
 
-        // we only "require" it here so that the revert reason on invalid signature will be of "VerifyingPaymaster", and not "ECDSA"
-        require(
-            signature.length == 65,
-            "BTPM: invalid signature length in paymasterAndData"
-        );
-
+        // gas = gasleft();
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:451]: should be removed

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 11]
```diff
@@ -443,43 +435,30 @@ contract BiconomyTokenPaymaster is
...
             validUntil,
             validAfter,
             feeToken,
-            oracleAggregator,
             exchangeRate,
             priceMarkup
         ).toEthSignedMessageHash();
-
-        context = "";
+        // console.log("gas used for getHash: %s", gas - gasleft());
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:461]: should be removed

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 12]
```diff
@@ -490,7 +469,7 @@ contract BiconomyTokenPaymaster is
             );
         }
 
-        address account = userOp.getSender();
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:493]: isn't `getSender()` more efficient?

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 13]
```diff
@@ -499,22 +478,17 @@ contract BiconomyTokenPaymaster is
...
-        );
         require(priceMarkup <= 2e6, "BTPM: price markup percentage too high");
-        require(priceMarkup >= 1e6, "BTPM: price markup percentage too low");
         require(
             IERC20(feeToken).balanceOf(account) >=
                 ((tokenRequiredPreFund * priceMarkup) / PRICE_DENOMINATOR),
             "BTPM: account does not have enough token balance"
         );
 
+        // gas = gasleft();
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:488]: should be removed

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 14]
```diff
@@ -538,36 +512,42 @@ contract BiconomyTokenPaymaster is
...
-            oracleAggregator != address(NATIVE_ADDRESS) &&
-            oracleAggregator != address(0)
+            priceSource == ExchangeRateSource.ORACLE_BASED 
         ) {
-            uint256 result = exchangePrice(address(feeToken), oracleAggregator);
-            if (result != 0) effectiveExchangeRate = result;
+            effectiveExchangeRate = getTokenValueOfOneNativeToken(address(feeToken));
+            // uint256 result = getTokenValueOfOneNativeToken(address(feeToken));
+            // Review
+            // if (result != 0) effectiveExchangeRate = result;
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:550]: should be removed

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 15]
```diff
@@ -590,12 +570,12 @@ contract BiconomyTokenPaymaster is
                 account,
                 address(feeToken),
                 charge,
-                oracleAggregator,
                 priceMarkup,
                 userOpHash,
                 effectiveExchangeRate,
                 priceSource
             );
+            // emit UserOperationSponsored(account, address(feeToken), charge, actualGasCost);
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:578]: is this event necessary?

File: contracts/token/BiconomyTokenPaymaster.sol [hunk 16]
```diff
@@ -612,6 +592,7 @@ contract BiconomyTokenPaymaster is
                 // Do nothing else here to not revert the whole bundle and harm reputation
             }
         }
+        // console.log("gas used for postop: %s", gas - gasleft());
```
Comment (ankurdubey521) [inline contracts/token/BiconomyTokenPaymaster.sol:595]: should be removed

**Comment by ankurdubey521** [comment]
Let me know once the `try {} catch {}` changes are implemented for Oracle Aggregator

File: contracts/token/oracles/OracleAggregator.sol (hunk 1 above)
Comment (livingrockrises) [reply contracts/token/oracles/OracleAggregator.sol:21]: ok

File: contracts/token/oracles/OracleAggregator.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/token/oracles/OracleAggregator.sol:54]: done

File: contracts/token/oracles/OracleAggregator.sol (hunk 3 above)
Comment (livingrockrises) [reply contracts/token/oracles/OracleAggregator.sol:99]: done

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 4 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol]: fixed

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 5 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol:131]: done

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 13 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol]: fixed

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 14 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol]: done

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 6 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol:119]: ok

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 9 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol]: done

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 11 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol]: done

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 15 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol]: removed

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 8 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol:390]: done

File: contracts/token/BiconomyTokenPaymaster.sol (hunk 7 above)
Comment (livingrockrises) [reply contracts/token/BiconomyTokenPaymaster.sol:319]: ok done

File: contracts/deployer/Create3.sol [hunk 17]
```diff
@@ -2,35 +2,34 @@
...
-      0x03  0x37  0x37                  CALLDATACOPY
-      0x04  0x36  0x36                  CALLDATASIZE    cds
-      0x05  0x3d  0x3d                  RETURNDATASIZE  0 cds
-      0x06  0x34  0x34                  CALLVALUE       val 0 cds
-      0x07  0xf0  0xf0                  CREATE          addr
-  */
-
-    bytes internal constant PROXY_CHILD_BYTECODE =
-        hex"67_36_3d_3d_37_36_3d_34_f0_3d_52_60_08_60_18_f3";
+    bytes internal constant PROXY_CHILD_BYTECODE = hex"67363d3d37363d34f03d5260086018f3";
```
Comment (tomarsachin2271) [inline contracts/deployer/Create3.sol:32]: Is this intentional or a part of formatting? Coz the value is also changed.

File: contracts/interfaces/paymasters/IBiconomyTokenPaymaster.sol [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        uint32 priceMarkup,
+        bytes32 userOpHash,
+        uint128 exchangeRate,
+        ExchangeRateSource priceSource
+    );
+
+    /**
+     * Notify in case paymaster failed to withdraw tokens from sender
+     */
+    event TokenPaymentDue(address indexed token, address indexed account, uint256 indexed charge);
```
Comment (tomarsachin2271) [inline contracts/interfaces/paymasters/IBiconomyTokenPaymaster.sol:53]: Do we need this TokenPaymentDue now?

File: contracts/sponsorship/PaymasterHelpers.sol [hunk 19]
```diff
@@ -29,7 +29,10 @@ library PaymasterHelpers {
      * @param data PaymasterData passed
      */
     function paymasterContext(
-        UserOperation calldata /** op */,
+        UserOperation calldata,
```
Comment (tomarsachin2271) [inline contracts/sponsorship/PaymasterHelpers.sol:32]: This formatting can be reverted.

File: contracts/sponsorship/SponsorshipPaymaster.sol [hunk 20]
```diff
@@ -247,7 +226,10 @@ contract SponsorshipPaymaster is
      * @param actualGasCost amount to be paid to the entry point in wei
      */
     function _postOp(
-        PostOpMode /** mode */,
+        PostOpMode,
+        /**
+         * mode
```
Comment (tomarsachin2271) [inline contracts/sponsorship/SponsorshipPaymaster.sol:231]: This formatting can be reverted.

File: contracts/sponsorship/PaymasterHelpers.sol (hunk 19 above)
Comment (livingrockrises) [reply contracts/sponsorship/PaymasterHelpers.sol:32]: done

File: contracts/sponsorship/SponsorshipPaymaster.sol (hunk 20 above)
Comment (livingrockrises) [reply contracts/sponsorship/SponsorshipPaymaster.sol:231]: done

File: contracts/deployer/Create3.sol (hunk 17 above)
Comment (livingrockrises) [reply contracts/deployer/Create3.sol:32]: both format works in solidity. value is same

**Review by tomarsachin2271 — APPROVED** [review]

**Comment by ankurdubey521** [comment]
LGTM. We should also move to 0.8.23, can address that in a separate PR

**Pull request merged**